File: presto-common/src/main/java/com/facebook/presto/common/Subfield.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.presto.common;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static java.util.Objects.requireNonNull;

public class Subfield
{
    public interface PathElement
    {
        boolean isSubscript();
    }

    public static final class AllSubscripts
            implements PathElement
    {
        private static final AllSubscripts ALL_SUBSCRIPTS = new AllSubscripts();

        private AllSubscripts() {}

        public static AllSubscripts getInstance()
        {
            return ALL_SUBSCRIPTS;
        }

        @Override
        public boolean isSubscript()
        {
            return true;
        }

        @Override
        public String toString()
        {
            return "[*]";
        }
    }

    public static final class NestedField
            implements PathElement
    {
        private final String name;

        public NestedField(String name)
        {
            this.name = requireNonNull(name, "name is null");
        }

        public String getName()
        {
            return name;
        }

        @Override
        public boolean equals(Object o)
        {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            NestedField that = (NestedField) o;
            return Objects.equals(name, that.name);
        }

        @Override
        public int hashCode()
        {
            return Objects.hash(name);
        }

        @Override
        public String toString()
        {
            return "." + name;
        }

        @Override
        public boolean isSubscript()
        {
            return false;
        }
    }

    public static final class LongSubscript
            implements PathElement
    {
        private final long index;

        public LongSubscript(long index)
        {
            this.index = index;
        }

        public long getIndex()
        {
            return index;
        }

        @Override
        public boolean equals(Object o)
        {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            LongSubscript that = (LongSubscript) o;
            return index == that.index;
        }

        @Override
        public int hashCode()
        {
            return Objects.hash(index);
        }

        @Override
        public String toString()
        {
            return "[" + index + "]";
        }

        @Override
        public boolean isSubscript()
        {
            return true;
        }
    }

    public static final class StringSubscript
            implements PathElement
    {
        private final String index;

        public StringSubscript(String index)
        {
            this.index = requireNonNull(index, "index is null");
        }

        public String getIndex()
        {
            return index;
        }

        @Override
        public boolean equals(Object o)
        {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            StringSubscript that = (StringSubscript) o;
            return Objects.equals(index, that.index);
        }

        @Override
        public int hashCode()
        {
            return Objects.hash(index);
        }

        @Override
        public String toString()
        {
            return "[\"" + index.replace("\"", "\\\"") + "\"]";
        }

        @Override
        public boolean isSubscript()
        {
            return true;
        }
    }

    private final String name;
    private final List<PathElement> path;

    public static PathElement allSubscripts()
    {
        return AllSubscripts.getInstance();
    }

    @JsonCreator
    public Subfield(String path)
    {
        requireNonNull(path, "path is null");

        SubfieldTokenizer tokenizer = new SubfieldTokenizer(path);
        checkArgument(tokenizer.hasNext(), "Column name is missing: " + path);

        PathElement firstElement = tokenizer.next();
        checkArgument(firstElement instanceof NestedField, "Subfield path must start with a name: " + path);

        this.name = ((NestedField) firstElement).getName();

        List<PathElement> pathElements = new ArrayList<>();
        tokenizer.forEachRemaining(pathElements::add);
        this.path = Collections.unmodifiableList(pathElements);
    }

    private static void checkArgument(boolean expression, String errorMessage)
    {
        if (!expression) {
            throw new IllegalArgumentException(errorMessage);
        }
    }

    public Subfield(String name, List<PathElement> path)
    {
        this.name = requireNonNull(name, "name is null");
        this.path = requireNonNull(path, "path is null");
    }

    public String getRootName()
    {
        return name;
    }

    public List<PathElement> getPath()
    {
        return path;
    }

    public boolean isPrefix(Subfield other)
    {
        if (!other.name.equals(name)) {
            return false;
        }

        if (path.size() < other.path.size()) {
            return Objects.equals(path, other.path.subList(0, path.size()));
        }

        return false;
    }

    public Subfield tail(String name)
    {
        if (path.isEmpty()) {
            throw new IllegalStateException("path is empty");
        }

        return new Subfield(name, path.subList(1, path.size()));
    }

    @JsonValue
    public String serialize()
    {
        return name + path.stream()
                .map(PathElement::toString)
                .collect(Collectors.joining());
    }

    @Override
    public String toString()
    {
        return serialize();
    }

    @Override
    public boolean equals(Object o)
    {
        if (this == o) {
            return true;
        }

        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        Subfield other = (Subfield) o;
        return Objects.equals(name, other.name) &&
                Objects.equals(path, other.path);
    }

    @Override
    public int hashCode()
    {
        return Objects.hash(name, path);
    }
}

File: velox/type/Subfield.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "velox/type/Subfield.h"
#include <boost/algorithm/string/replace.hpp>
#include <folly/container/F14Map.h>
#include "velox/type/Tokenizer.h"

namespace facebook::velox::common {

namespace {
const auto& subfieldKindNames() {
  static const folly::F14FastMap<SubfieldKind, std::string_view> kNames = {
      {SubfieldKind::kAllSubscripts, "AllSubscripts"},
      {SubfieldKind::kNestedField, "NestedField"},
      {SubfieldKind::kStringSubscript, "StringSubscript"},
      {SubfieldKind::kLongSubscript, "LongSubscript"},
  };

  return kNames;
}
} // namespace

VELOX_DEFINE_ENUM_NAME(SubfieldKind, subfieldKindNames)

Subfield::Subfield(
    const std::string& path,
    std::shared_ptr<const Separators> separators) {
  Tokenizer tokenizer{path, std::move(separators)};
  VELOX_CHECK(tokenizer.hasNext(), "Column name is missing: {}", path);

  auto firstElement = tokenizer.next();
  VELOX_CHECK_EQ(
      firstElement->kind(),
      SubfieldKind::kNestedField,
      "Subfield path must start with a name: {}",
      path);

  std::vector<std::unique_ptr<PathElement>> pathElements;
  pathElements.push_back(std::move(firstElement));
  while (tokenizer.hasNext()) {
    pathElements.push_back(tokenizer.next());
  }
  path_ = std::move(pathElements);
}

Subfield::Subfield(std::vector<std::unique_ptr<Subfield::PathElement>>&& path)
    : path_(std::move(path)) {
  VELOX_CHECK_GE(path_.size(), 1);
  VELOX_CHECK_EQ(
      path_[0]->kind(),
      SubfieldKind::kNestedField,
      "Subfield path must start with a name");
}

Subfield Subfield::clone() const {
  Subfield subfield;
  subfield.path_.reserve(path_.size());
  for (auto& element : path_) {
    subfield.path_.push_back(element->clone());
  }
  return subfield;
}

bool Subfield::isPrefix(const Subfield& other) const {
  if (path_.size() < other.path_.size()) {
    for (size_t i = 0; i < path_.size(); ++i) {
      if (!(*path_[i].get() == *other.path_[i].get())) {
        return false;
      }
    }
    return true;
  }
  return false;
}

std::string Subfield::toString() const {
  if (!valid()) {
    return "";
  }
  std::ostringstream out;
  out << static_cast<const NestedField*>(path_[0].get())->name();
  for (size_t i = 1; i < path_.size(); i++) {
    out << path_[i]->toString();
  }
  return out.str();
}

bool Subfield::operator==(const Subfield& other) const {
  if (this == &other) {
    return true;
  }

  if (path_.size() != other.path_.size()) {
    return false;
  }
  for (size_t i = 0; i < path_.size(); ++i) {
    if (!(*path_[i].get() == *other.path_[i].get())) {
      return false;
    }
  }
  return true;
}

size_t Subfield::hash() const {
  size_t result = 1;
  for (size_t i = 0; i < path_.size(); ++i) {
    result = result * 31 + path_[i]->hash();
  }
  return result;
}

std::string Subfield::StringSubscript::toString() const {
  return "[\"" + boost::replace_all_copy(index_, "\"", "\\\"") + "\"]";
}

} // namespace facebook::velox::common
